File: Program's_Contributed_By_Contributors/C++ Programs/Array/binarySearch/binarySearch.cpp
#include <bits/stdc++.h>
using namespace std;
// this generate function generates a random array of given size and range
void generate(vector<int> &a, int size, int range)
{
    for (int i = 0; i < size; i++)
    {
        a.push_back(rand() % range);
    }
}

// actual binary search program 
int binary_search(vector<int> &t, int x)
{
    int p = 0;
    int r = t.size();
    int mid;
    int low = p;
    int high = max(p, r + 1), count = 0;
    while ((low < high))
    {
        mid = (low + high) / 2;
        if ((x <= t[mid]))
        {
            ++count;
            high = mid;
        }
        else
        {
            ++count;
            low = mid + 1;
        }
    }
    cout << "c1: " << count << endl;
    cout << "c2: " << log(r) << endl;
    cout << "Const. of proportionality: " << count / (log(r)) << endl; // shows the ratio between: (actual number of comparisions) / (expected number of comparisions)
    if (t[high] == x)
        return high;
    return -1;
}

int main()
{
    vector<int> array;
    int size, range, search;
    cout << "Enter the size of array: " << endl;
    cin >> size;
    cout << "Enter the range of random numebrs: " << endl;
    cin >> range;
    generate(array, size, range);
    sort(array.begin(), array.end());
    cout << "Sorted array: " << endl;
    for (int i = 0; i < array.size(); i++)
        cout << array[i] << " ";

    cout << "\nEnter the element to binary search: " << endl;
    cin >> search;
    int found = binary_search(array, search);
    if (found > -1)
    {
        cout << "Element FOUND at index: " << found << endl;
    }
    else
    {
        cout << "Element NOT FOUND" << endl;
    }
    return 0;
}

File: Program's_Contributed_By_Contributors/Java_Programs/Array/BinarySearch.java
import java.util.*;

public class BinarySearch {
    public static int binarysearch(int numbers[], int key){
        int start = 0;
        int end = numbers.length-1;
        while(start<=end){
            int mid = (start+end)/2;
            if(numbers[mid]==key){
                return mid;
            }
            if(numbers[mid]<key){
                start = mid + 1;
            }
            else{
                end = mid - 1;
            }
        }
        return -1;
    }

    public static void main(String args[]){
        int numbers[] = {1,2,5,3,8,6,9,10,33,66};
        int key = 10;
        System.out.println("Index of key is: "+ binarysearch(numbers, key));

    }
    
}
